### PR TITLE
Audio: TDFB: Optimize the circular buffer usage

### PR DESCRIPTION
**Pull request by andrula-song**
Optimize the tdfb_fir_sx functions by remove the duplicated usage of circular buffer.
the performance I got by xt-testbench is here:

Format		original(mcps)		optimized(mcps)	saving mcps
16/16		17.42			17.23			0.19
24/32		17.59			17.32			0.27
32/32		17.30			17.12			0.18

### CODE REVIEW

File: src/audio/tdfb/tdfb_hifi3.c [hunk 1]
```diff
@@ -38,54 +38,68 @@ void tdfb_fir_s16(struct tdfb_comp_data *cd, struct input_stream_buffer *bsource
...
+
+				/* Get filter instance */
+				f = &cd->fir[i];
+				shift = -f->out_shift;
+
+				/* Compute FIR and mix as Q5.27*/
+				fir_core_setup_circular(f);
+				fir_32x16_2x_hifi3(f, cd->in[is], cd->in[is2], &y0, &y1,
+						   shift);
+				for (k = 0; k < out_nch; k++) {
```
Comment (lyakh) [inline src/audio/tdfb/tdfb_hifi3.c:79]: these loops could also be optimised a bit further by using some form of `ffs()`

**Comment by btian1** [comment]
shall we expect same code change on generic c code? c code should be always same with optimized code in case one day hifi version have bugs.

**Comment by andrula-song** [comment]
thanks for your comment, but actually I just drafted this PR for experts evaluation, since there is very little improvement, I think it is meaningless.

**Review by singalsu — APPROVED** [review]
Looks good to me plus, the small saving is nice. Thank you!

**Comment by singalsu** [comment]
You should add to commit message "Fixes #5273"

File: src/include/sof/audio/tdfb/tdfb_comp.h [hunk 2]
```diff
@@ -24,7 +24,7 @@
 #define TDFB_GENERIC	0
 #define TDFB_HIFIEP	1
 #define TDFB_HIFI3	0
-#elif XCHAL_HAVE_HIFI3 == 1
```
Comment (kv2019i) [inline src/include/sof/audio/tdfb/tdfb_comp.h:27]: This would better be in a separate commit, not related to the performance optimizations.

**Review by kv2019i — COMMENTED** [review]
@andrula-song Can you split the hifi4 change to a different commit (can be in same PR)? Otherwise looks good to go.

**Review by kv2019i — APPROVED** [review]

**Comment by kv2019i** [comment]
The fails in https://sof-ci.01.org/sofpr/PR7700/build10102/devicetest/index.html are worrying, but I can't figure out how it could be related to this PR. I'll proceed with merge based on earlier CI results for this PR and CI results for other PRs today.

**Comment by fredoh9** [comment]
This cause DSP exception on boot and runtime pm test.
After I reverted this, there is no DSP exception. Tested with MTL NOCODEC.

**Comment by keqiaozhang** [comment]
This PR brings a regression https://github.com/thesofproject/linux/issues/4442
